pkg/enterpriseclient/auth.go

```diff
@@ -191,26 +194,124 @@ func homeDir() string {
 	return os.Getenv("USERPROFILE")
 }
 
+type SigBlock struct {
+	Nonce     []byte
+	Timestamp []byte
+	Signature []byte
+}
+
 // gets the (base64 encoded) signature and fingerprint for a given key and data
-func sigAndFingerprint(privateKey *ecdsa.PrivateKey, data []byte) (string, string, error) {
+// returns the signature with a nonce and timestamp, the signature of the data alone, the fingerprint, and error
+func sigAndFingerprint(privateKey *ecdsa.PrivateKey, data []byte) (string, string, string, error) {
+	// generate a timestamp and nonce
+	nonce := make([]byte, 512/8)        // 512 bits
+	_, _ = rand.Read(nonce)             // returns len, nil - neither of which we need
+	ts, err := time.Now().MarshalText() // marshals to RFC3339Nano
+	if err != nil {
+		return "", "", "", errors.Wrap(err, "failed to get timestamp")
+	}
+
 	// hash the body and sign the hash
 	// store the signature in an ecSig struct and marshal it with the ssh wire format
 	contentSha := sha512.Sum512(data)
 	var ecSig struct {
 		R *big.Int
 		S *big.Int
 	}
-	var err error
 	ecSig.R, ecSig.S, err = ecdsa.Sign(rand.Reader, privateKey, contentSha[:])
 	if err != nil {
-		return "", "", errors.Wrap(err, "failed to sign content sha")
+		return "", "", "", errors.Wrap(err, "failed to sign content sha")
 	}
 	signatureString := base64.StdEncoding.EncodeToString(ssh.Marshal(ecSig))
 
+	// do the same for the data combined with the timestamp and nonce
+	contentSha = sha512.Sum512(combineTsNonceData(ts, nonce, data))
+	ecSig.R, ecSig.S, err = ecdsa.Sign(rand.Reader, privateKey, contentSha[:])
+	if err != nil {
+		return "", "", "", errors.Wrap(err, "failed to sign content sha")
+	}
+	tsSig := ssh.Marshal(ecSig)
+
 	// include the public key fingerprint as a hint to the server
 	fingerprint, err := getFingerprint(&privateKey.PublicKey)
 	if err != nil {
-		return "", "", errors.Wrap(err, "failed to get public key fingerprint")
+		return "", "", "", errors.Wrap(err, "failed to get public key fingerprint")
+	}
+
+	sigBlock := SigBlock{
+		Timestamp: ts,
+		Nonce:     nonce,
+		Signature: tsSig,
 	}
-	return signatureString, fingerprint, nil
+	sigBlockBytes, err := json.Marshal(sigBlock)
+	if err != nil {
+		return "", "", "", errors.Wrap(err, "failed to marshal signature block")
+	}
+	sigBlockString := base64.StdEncoding.EncodeToString(sigBlockBytes)
+
+	return sigBlockString, signatureString, fingerprint, nil
+}
+
+func combineTsNonceData(ts, nonce, data []byte) []byte {
+	tsBytes := []byte{}
+	tsBytes = append(tsBytes, nonce...)
+	tsBytes = append(tsBytes, ts...)
+	tsBytes = append(tsBytes, data...)
+	return tsBytes
+}
+
+// ValidatePayload checks that the payload was signed by the provided public key
```

Comment:
signed by the public key? don't you mean the private key? or maybe i don't get the phrasing

Reply:
yeah, bad phrasing
"signed by the private key associated with the provided public key" or something